modules/metastorage/src/main/java/org/apache/ignite/metastorage/internal/watch/WatchAggregator.java

```diff
@@ -0,0 +1,254 @@
+                       if (watch.keyCriterion.contains(evt.oldEntry().key()) &&
+                           (watch.revision < 0 || evt.oldEntry().revision() >= watch.revision))
+                           filteredEvts.add(evt);
+                    }
+
+                    if (!filteredEvts.isEmpty()) {
+                        if (!watch.lsnr.onUpdate(filteredEvts))
+                            watchIt.remove();
+                    }
+                }
+                return true;
+            }
+
+            @Override public void onError(@NotNull Throwable e) {
+                watches.values().forEach(w -> w.lsnr.onError(e));
+            }
+        };
+    }
+
+    /** */
+    private static interface KeyCriterion {
+        /** */
+        public IgniteBiTuple<Key, Key> toRange();
+
+        /** */
+        public boolean contains(Key key);
+    }
+
+    /** */
+    private static class ExactCriterion implements KeyCriterion {
+        /** */
+        private final Key key;
+
+        /** */
+        public ExactCriterion(Key key) {
+            this.key = key;
+        }
+
+        /** */
+        @Override public IgniteBiTuple<Key, Key> toRange() {
+            return new IgniteBiTuple<>(key, key);
+        }
+
+        /** */
+        @Override public boolean contains(Key key) {
+            return this.key.equals(key);
+        }
+
+    }
+
+    /** */
+    private static class RangeCriterion implements KeyCriterion {
+        /** */
+        private final Key from;
+
+        /** */
+        private final Key to;
+
+        /** */
+        public RangeCriterion(Key from, Key to) {
+            this.from = from;
+            this.to = to;
+        }
+
+        /** */
+        @Override public IgniteBiTuple<Key, Key> toRange() {
+            return new IgniteBiTuple<>(from, to);
+        }
+
+        /** */
+        @Override public boolean contains(Key key) {
+            return key.compareTo(from) >= 0 && key.compareTo(to) < 0;
+        }
+    }
+
+    /** */
+    private static class CollectionCriterion implements KeyCriterion {
+        /** */
+        private final Collection<Key> keys;
+
+        /** */
+        public CollectionCriterion(Collection<Key> keys) {
+            this.keys = keys;
+        }
+
+        /** */
+        @Override public IgniteBiTuple<Key, Key> toRange() {
+            return new IgniteBiTuple<>(Collections.min(keys), Collections.max(keys));
+        }
+
+        /** */
+        @Override public boolean contains(Key key) {
+            return keys.contains(key);
+        }
+    }
+
+    /** */
+    private static class PrefixCriterion implements KeyCriterion {
+        /** */
+        private final Key prefixKey;
+
+        /** */
+        public PrefixCriterion(Key prefixKey) {
+            this.prefixKey = prefixKey;
+        }
+
+        /** */
+        @Override public IgniteBiTuple<Key, Key> toRange() {
+            var bytes = Arrays.copyOf(prefixKey.bytes(), prefixKey.bytes().length);
```

Comment:
Seems that we should use org.apache.ignite.metastorage.common.Key convertation here: `s.getBytes(StandardCharsets.UTF_8)`

Reply:
Not sure, that I understand - prefixKey is a Key instance already. So, here I just want to copy its content as is.

Reply:
Yep you are right.